ludwig/trainers/trainer.py

```diff
@@ -1189,8 +1183,26 @@ def _train_loop(
                     if self.is_coordinator():
                         progress_tracker.save(os.path.join(save_path, TRAINING_PROGRESS_TRACKER_FILE_NAME))
 
+                # If this was the last batch, then increment the epoch counter and invoke the `on_epoch_end` callback.
+                if not has_called_on_epoch_end and batcher.last_batch():
```

Comment:
I'm not a huge fan of the duplicate code blocks. If we remove the early return below, this block should be unnecessary, right?

Reply:
+1